File: CPP/container.cpp
#include "container.h"
#include "baseMatrix.h"
#include "matrix.h"
#include "diagonalMatrix.h"
#include "lowerTriangularMatrix.h"
#include <cstdio>
#include <iostream>

// Base constructor.
Container::Container() {
    size_ = 0;
}

// Determined size constructor.
Container::Container(int size) {
    size_ = size;
}

// One matrix add.
void Container::Add(int index, FILE *input) {
    // Creating new matrix.
    BaseMatrix *bm;
    int type;
    fscanf(input, "%d\n", &type);
    int size;
    fscanf(input, "%d\n", &size);

    // Selecting type.
    if (type == 0) {
        bm = new Matrix(size);
    } else if (type == 1) {
        bm = new DiagonalMatrix(size);
    } else if (type == 2) {
        bm = new LowerTriangularMatrix(size);
    } else {
        return;
    }

    // Reading and adding.
    bm->Read(input);
    matrices[index] = bm;
}

// File input.
void Container::In(FILE *input) {
    for (int i = 0; i < size_; ++i) {
        Add(i, input);
    }
}

// Random input.
void Container::RandomIn(int number) {
    for (int i = 0; i < number; ++i) {
        // Random type.
        int random_type = rand() % 3;

        if (random_type == 0) {
            matrices[i] = new Matrix(0);
        } else if (random_type == 1) {
            matrices[i] = new DiagonalMatrix(0);
        } else {
            matrices[i] = new LowerTriangularMatrix(0);
        }

        // Random matrix.
        matrices[i]->RandomRead();
    }
}

// Outputting all the matrices.
void Container::Out(FILE *output) {
    fprintf(output, "Container size: %d\n", size_);
    fprintf(output, "%s\n", "----------------------------------------");

    for (int i = 0; i < size_; ++i) {
        fprintf(output, "%s\n", "\n");
        matrices[i]->Out(output);
        fprintf(output, "%s\n", "\n");
    }

    fprintf(output, "%s\n", "----------------------------------------");
}

// Sorting all the matrices.
void Container::Sort() {
    for (int i = 0; i < size_ - 1; i++) {
        // Getting current minimum index.
        int min_ind = i;
        for (int j = i + 1; j < size_; j++) {
            if (matrices[j]->getAverage() < matrices[min_ind]->getAverage())
                min_ind = j;
        }

        // Changing.
        BaseMatrix *new_matrix = matrices[min_ind];
        matrices[min_ind] = matrices[i];
        matrices[i] = new_matrix;
    }
}

File: Python/container.py
import random
from matrices.matrix import Matrix
from matrices.diagonalMatrix import DiagonalMatrix
from matrices.lowerTriangularMatrix import LowerTriangularMatrix


# Container class.
class Container:
    # Container size.
    __size__: int
    # Container itself.
    __container__: list

    # Container constructor.
    def __init__(self, size: int):
        self.__size__ = size
        self.__container__ = []

    # Reading from file.
    def file_in(self, lines: list):
        counter = 0
        for i in range(0, self.__size__):
            matrix_type = int(lines[counter])
            counter += 1
            num = int(lines[counter])
            counter += 1

            if matrix_type == 0:
                m = Matrix(num)
                m.read(lines[counter])
                self.__container__.append(m)
                counter += 1
            elif matrix_type == 1:
                dm = DiagonalMatrix(num)
                dm.read(lines[counter])
                self.__container__.append(dm)
                counter += 1
            else:
                ltm = LowerTriangularMatrix(num)
                ltm.read(lines[counter])
                self.__container__.append(ltm)
                counter += 1

    # Creating randomly.
    def random_in(self):
        for i in range(0, self.__size__):
            matrix_type = random.randint(0, 3)

            if matrix_type == 0:
                m = Matrix(0)
                m.random_read()
                self.__container__.append(m)
            elif matrix_type == 1:
                dm = DiagonalMatrix(0)
                dm.random_read()
                self.__container__.append(dm)
            else:
                ltm = LowerTriangularMatrix(0)
                ltm.random_read()
                self.__container__.append(ltm)

    # Outputting.
    def out(self) -> str:
        ans = 'Number of elements: ' + str(self.__size__) + '\n'
        ans += '-' * 20 + '\n'

        for item in self.__container__:
            ans += '\n\n'
            ans += item.out()
            ans += '\n\n'

        ans += '-' * 20 + '\n'
        return ans

    # Sorting.
    def sorting(self):
        for i in range(0, self.__size__ - 1):
            # Current minimum.
            min_ind = i

            for j in range(i + 1, self.__size__):
                if self.__container__[j].get_average() < \
                        self.__container__[min_ind].get_average():
                    min_ind = j

            # Swapping.
            self.__container__[min_ind], self.__container__[i] = \
                self.__container__[i], self.__container__[min_ind]
